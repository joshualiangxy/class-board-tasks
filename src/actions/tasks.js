import firebase from '../firebase/firebase';

export const addPersonalTask = task => ({ type: 'ADD_PERSONAL_TASK', task });

export const startAddPersonalTask = task => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const id = task.id;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .set(task)
      .then(() => dispatch(addPersonalTask(task)));
  };
};

export const removePersonalTask = id => ({ type: 'REMOVE_PERSONAL_TASK', id });

export const startRemovePersonalTask = id => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .delete()
      .then(() => dispatch(removePersonalTask(id)));
  };
};

export const editPersonalTask = (id, updates) => ({
  type: 'EDIT_PERSONAL_TASK',
  id,
  updates
});

export const startEditPersonalTask = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .set(updates)
      .then(dispatch(editPersonalTask(id, updates)));
  };
};

export const setTasks = tasks => ({ type: 'SET_TASK', tasks });

export const startSetTasks = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const tasks = [];
    const userRef = firebase.firestore().collection('users').doc(uid);
    const tasksRef = userRef.collection('tasks');

    return tasksRef
      .get()
      .then(querySnapshot =>
        querySnapshot.forEach(queryDocSnapshot =>
          tasks.push({ id: queryDocSnapshot.id, ...queryDocSnapshot.data() })
        )
      )
      .then(() => {
        userRef.get().then(userDocSnapshot => {
          const groupIds = userDocSnapshot.get('groups') || [];
          const groupsPromises = [];

          groupIds.forEach(gid => {
            const groupPromises = [];
            const groupRef = firebase.firestore().collection('groups').doc(gid);

            groupPromises.push(
              groupRef
                .get()
                .then(groupDocSnapshot => groupDocSnapshot.get('name'))
            );
            groupPromises.push(groupRef.collection('tasks').get());

            groupsPromises.push(
              Promise.all(groupPromises).then(
                ([groupName, queryTaskCollectionSnapshot]) =>
                  queryTaskCollectionSnapshot.forEach(queryTaskDocSnapshot =>
                    tasks.push({
                      groupName,
                      gid,
                      id: queryTaskDocSnapshot.id,
                      ...queryTaskDocSnapshot.data()
                    })
                  )
              )
            );

            return Promise.all(groupPromises);
          });
        });
      })
      .then(() => dispatch(setTasks(tasks)));
  };
};

export const toggleCompletedPersonal = (id, completedState) => ({
  type: 'TOGGLE_COMPLETED_PERSONAL',
  id,
  completedState
});

export const startToggleCompletedPersonal = (id, completedState) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .update({ completed: !completedState })
      .then(() => toggleCompletedPersonal(id, completedState));
  };
};

export const removeTaskData = () => ({ type: 'REMOVE_TASK_DATA' });
