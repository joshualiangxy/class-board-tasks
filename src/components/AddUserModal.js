import React, { useState } from 'react';
import Modal from 'react-modal';
import firebase from '../firebase/firebase';

// Note: This still works off uid for now. This also does NOT work yet.
const AddUserModal = ({ isOpen, onRequestClose, group }) => {
  const addNewUser = (userEmail, group) => {
    const uid = userEmail; // TODO: Change this when email-uid are linked!

    // Both of these references are DocumentReference
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(group);

    // Gets the data from a user.
    userRef.get().then(doc => {
      const data = doc.data();
      // Write displayName, studentNum, and UID into a new document
      if (typeof data === 'undefined') {
        // Most likely a user that doesn't exist.
        setError('No such user found');
      } else {
        groupRef
          .collection('users')
          .doc(uid) // Create document for this user's UID
          .set({
            studentNum: data.studentNum,
            displayName: data.displayName,
            uid: uid,
            admin: false
          })
          .then(() => {
            setError('User added!');
          })
          .catch(error => {
            setError('Something went wrong.');
          });
      }
      // Defaults to non-admin.
    });
  };

  // TODO: When calling this modal, pass in the group UID as a prop
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');
  // const [usersAdded, setUsersAdded] = useState([]); // For listing down people that have been added?

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim();
    if (!submittedEmail) {
      setError('Please enter a group name');
    } else {
      addNewUser(submittedEmail, group);
      // TODO: Catch user not found
      // Does not close after submitting, so that the user can continue to add more users

      // setUserEmail(''); // Also resets the field.
      // setError('Added!');
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="User ID (required)"
          value={userEmail}
          onChange={onUserEmailChange}
          autoFocus
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

export default AddUserModal;
