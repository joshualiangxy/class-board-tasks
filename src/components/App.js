import React from 'react';
import '../styles/App.css';

const App = () => (
  <div>
    <h1>This is my app</h1>
  </div>
);

export default App;

