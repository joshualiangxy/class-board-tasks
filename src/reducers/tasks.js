const tasksReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_TASK':
      return [...state, action.task];
    case 'REMOVE_TASK':
      return state.filter(task => task.id !== action.id);
    case 'EDIT_TASK':
      return state.map(task =>
        task.id === action.id
          ? {
              ...task,
              ...action.updates
            }
          : task
      );
    case 'SET_TASK':
      return action.tasks;
    default:
      return state;
  }
};

export default tasksReducer;
