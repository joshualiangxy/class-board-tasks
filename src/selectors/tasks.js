const getSortedTasks = (tasks, { text, sortBy, grouped }) => {
  const lowercaseText = text.toLowerCase();

  const processedTasks = tasks
    .filter(
      task =>
        task.title.toLowerCase().includes(lowercaseText) ||
        task.description.toLowerCase().includes(lowercaseText) ||
        task.module.toLowerCase().includes(lowercaseText)
    )
    .sort((taskOne, taskTwo) => {
      switch (sortBy) {
        case 'deadline':
          if (taskOne.deadline && taskTwo.deadline)
            return taskOne.deadline - taskTwo.deadline;
          else if (taskOne.deadline) return -1;
          else if (taskTwo.deadline) return 1;
          else return taskOne.title.localeCompare(taskTwo.title);
        case 'name':
          return taskOne.title.localeCompare(taskTwo.title);
      }
    });

  const sortedList = {};

  if (grouped) {
    processedTasks.forEach(task => {
      if (task.groupName) {
        const gid = task.gid;

        if (!sortedList[gid])
          sortedList[gid] = { groupName: task.groupName, tasks: [] };
        sortedList[gid].tasks.push(task);
      } else if (task.module) {
        const module = task.module;

        if (!sortedList[module])
          sortedList[module] = { groupName: task.module, tasks: [] };
        sortedList[module].tasks.push(task);
      } else {
        if (!sortedList.others)
          sortedList.others = { groupName: 'others', tasks: [] };
        sortedList.others.tasks.push(task);
      }
    });
  } else sortedList.others = { groupName: 'others', tasks: processedTasks };

  return sortedList;
};

export default getSortedTasks;
