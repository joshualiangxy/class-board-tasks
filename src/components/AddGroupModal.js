import React, { useState } from 'react';
import Modal from 'react-modal';
import { startNewGroup } from '../actions/groups';
import { connect } from 'react-redux';

// TODO: Add the ability to add other users into the group
const AddGroupModal = ({ isOpen, onRequestClose, startNewGroup }) => {
  const [groupName, setGroupName] = useState('');
  const [module, setModule] = useState('');
  const [error, setError] = useState('');

  const onGroupNameChange = e => setGroupName(e.target.value);

  const onModuleChange = e => {
    const module = e.target.value.toUpperCase();

    setModule(module);
  };

  const onCancel = () => {
    setError('');
    setGroupName('');
    setModule('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedName = groupName.trim();
    const submittedMod = module.trim();
    if (!submittedName) {
      setError('Please enter a group name');
    } else {
      setError('');
      setGroupName('');
      setModule('');
      startNewGroup(submittedName, submittedMod).then(() => onRequestClose());
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={onCancel}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Group name (required)"
          value={groupName}
          onChange={onGroupNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Module code/Class name (Optional)"
          value={module}
          onChange={onModuleChange}
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startNewGroup: (groupName, module) =>
    dispatch(startNewGroup(groupName, module))
});

export default connect(null, mapDispatchToProps)(AddGroupModal);
