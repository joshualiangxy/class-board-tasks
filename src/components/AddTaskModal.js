import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startAddPersonalTask, startAddGroupTask } from '../actions/tasks';
import TaskForm from './TaskForm';

export const AddTaskModal = ({
  startAddPersonalTask,
  startAddGroupTask,
  removeAdmin,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  addGroupTask,
  users
}) => {
  const submitTask = task =>
    gid
      ? startAddGroupTask(task, groupName)
          .then(isAdmin => {
            if (isAdmin) return addGroupTask(task);
            else return removeAdmin();
          })
          .then(() => onRequestClose())
      : startAddPersonalTask(task).then(() => onRequestClose());

  return (
    <Modal
      className="task-modal"
      isOpen={isOpen}
      contentLabel="Add Task"
      onRequestClose={onRequestClose}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <h1>Add Task</h1>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddPersonalTask: task => dispatch(startAddPersonalTask(task)),
  startAddGroupTask: (task, groupName) =>
    dispatch(startAddGroupTask(task, groupName))
});

export default connect(undefined, mapDispatchToProps)(AddTaskModal);
