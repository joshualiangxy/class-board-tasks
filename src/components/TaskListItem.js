import React, { useState, useEffect } from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import {
  startRemovePersonalTask,
  startRemoveGroupTask,
  startToggleCompletedPersonal,
  startToggleCompletedGroup
} from '../actions/tasks';
import EditTaskModal from './EditTaskModal';
import FileUploadForm from './FileUploadForm';
import { downloadFile } from '../actions/files';

export const TaskListItem = ({
  removeAdmin,
  uid,
  task,
  users,
  dashboard,
  startRemovePersonalTask,
  startRemoveGroupTask,
  startToggleCompletedPersonal,
  startToggleCompletedGroup,
  removeGroupTask,
  toggleGroupTaskComplete,
  removeUserFromTask,
  showGroup,
  admin,
  groupName,
  editGroupTask,
  downloadFile
}) => {
  const {
    id,
    title,
    description,
    module,
    deadline,
    completed: initialComplete,
    gid,
    namingConvention,
    uploadRequired
  } = task;
  const [userInvolved, setUserInvolved] = useState(
    initialComplete.hasOwnProperty(uid)
  );
  const [visible, setVisible] = useState(false);
  const [completed, setCompleted] = useState(
    !dashboard ? (userInvolved ? initialComplete[uid] : false) : initialComplete
  );
  const [open, setOpen] = useState(false);
  const now = moment().valueOf;
  const deadlineClass = deadline > now ? 'before-deadline' : 'after-deadline';

  useEffect(() => {
    if (!(dashboard || userInvolved)) setCompleted(false);
  }, [dashboard, userInvolved]);

  useEffect(() => {
    setUserInvolved(initialComplete.hasOwnProperty(uid));
  }, [initialComplete, uid]);

  const toggleVisibility = () => setVisible(!visible);

  const toggleCompleted = () => {
    if (!(userInvolved || dashboard)) return;

    if (gid) {
      startToggleCompletedGroup(id, gid, completed).then(userInvolved => {
        if (!userInvolved) {
          setUserInvolved(false);
          if (!dashboard) removeUserFromTask(id, uid);
          return;
        }
        if (!dashboard) toggleGroupTaskComplete(id);
      });
    } else startToggleCompletedPersonal(id, completed);
    setCompleted(!completed);
  };

  const onRemove = e => {
    e.stopPropagation();
    gid
      ? startRemoveGroupTask(gid, id).then(isAdmin => {
          if (isAdmin) return removeGroupTask(id);
          else return removeAdmin();
        })
      : startRemovePersonalTask(id);
  };

  const openEditTask = e => {
    e.stopPropagation();
    setOpen(true);
  };

  const onRequestClose = () => setOpen(false);

  const onDownload = e => {
    e.stopPropagation();
    downloadFile(gid, id).then(isAdmin => {
      if (!isAdmin) return removeAdmin();
    });
  };

  return (
    <div>
      <input
        type="checkbox"
        checked={completed}
        onChange={toggleCompleted}
        disabled={!(dashboard || userInvolved)}
      />
      <div onClick={toggleVisibility}>
        <h3>{title}</h3>
        {deadline && (
          <h3 className={deadlineClass}>
            {moment(deadline).format('Do MMM YYYY')}
          </h3>
        )}
        {visible && (
          <div>
            {showGroup && module && <h5>{module}</h5>}
            {description && <p>{description}</p>}
            {userInvolved && uploadRequired && (
              <div>
                <FileUploadForm
                  id={id}
                  gid={gid}
                  namingConvention={namingConvention}
                />
              </div>
            )}
            {uploadRequired && admin && (
              <button onClick={onDownload}>Download submissions</button>
            )}
            {!dashboard &&
              Object.keys(initialComplete)
                .sort((uidOne, uidTwo) => {
                  const userOne = users.find(user => user.uid === uidOne);
                  const userTwo = users.find(user => user.uid === uidTwo);

                  if (!(userOne && userTwo)) return 0;
                  return userOne.displayName.localeCompare(userTwo.displayName);
                })
                .map(id => (
                  <div key={id}>
                    {users.find(user => user.uid === id) && (
                      <div>
                        <input
                          type="checkbox"
                          checked={id === uid ? completed : initialComplete[id]}
                          htmlFor={id}
                          disabled={true}
                        />
                        <label id={id}>
                          {users.find(user => user.uid === id).displayName}
                        </label>
                      </div>
                    )}
                  </div>
                ))}
            {(admin || !gid) && (
              <button onClick={openEditTask}>Edit Task</button>
            )}
            {(admin || !gid) && <button onClick={onRemove}>Remove Task</button>}
          </div>
        )}
      </div>
      <EditTaskModal
        removeAdmin={removeAdmin}
        isOpen={open}
        onRequestClose={onRequestClose}
        gid={gid}
        groupModule={module}
        groupName={groupName}
        task={task}
        users={users}
        editGroupTask={editGroupTask}
      />
    </div>
  );
};

const mapStateToProps = state => ({ uid: state.auth.user.uid });

const mapDispatchToProps = dispatch => ({
  startRemovePersonalTask: id => dispatch(startRemovePersonalTask(id)),
  startRemoveGroupTask: (gid, id) => dispatch(startRemoveGroupTask(gid, id)),
  startToggleCompletedPersonal: (id, completedState) =>
    dispatch(startToggleCompletedPersonal(id, completedState)),
  startToggleCompletedGroup: (id, gid, completedState) =>
    dispatch(startToggleCompletedGroup(id, gid, completedState)),
  downloadFile: (gid, id) => dispatch(downloadFile(gid, id))
});

export default connect(mapStateToProps, mapDispatchToProps)(TaskListItem);
