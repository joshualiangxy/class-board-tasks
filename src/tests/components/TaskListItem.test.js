import React from 'react';
import { shallow } from 'enzyme';
import { TaskListItem } from '../../components/TaskListItem';
import tasks from '../fixtures/tasks';

let wrapper;
const startRemovePersonalTask = jest.fn();
const startRemoveGroupTask = jest.fn(() => Promise.resolve());
const removeGroupTask = jest.fn();
const startToggleCompletedPersonal = jest.fn();
const startToggleCompletedGroup = jest.fn(() => Promise.resolve());
const task = tasks[0];
const id = task.id;
const completed = task.completed;

beforeEach(() => {
  wrapper = shallow(
    <TaskListItem
      task={task}
      startRemovePersonalTask={startRemovePersonalTask}
      startRemoveGroupTask={startRemoveGroupTask}
      removeGroupTask={removeGroupTask}
      startToggleCompletedPersonal={startToggleCompletedPersonal}
      startToggleCompletedGroup={startToggleCompletedGroup}
      showGroup={false}
      dashboard={true}
    />
  );
});

describe('render', () => {
  it('should render TaskListItem', () => expect(wrapper).toMatchSnapshot());

  it('should render TaskListItem with alt task', () => {
    wrapper.setProps({ task: tasks[1] });

    expect(wrapper).toMatchSnapshot();
  });
});

describe('completed checkbox', () => {
  it('should toggle completed state on change', () => {
    expect(wrapper.find('input').prop('checked')).toBe(completed);

    wrapper.find('input').prop('onChange')();

    expect(wrapper.find('input').prop('checked')).toBe(!completed);

    expect(startToggleCompletedPersonal).toHaveBeenCalledTimes(1);
    expect(startToggleCompletedPersonal).toHaveBeenLastCalledWith(
      id,
      completed
    );
  });
});

describe('clickable div', () => {
  it('should toggle visibility of div', () => {
    wrapper.find('div').at(1).prop('onClick')();

    expect(wrapper).toMatchSnapshot();
  });
});

describe('edit task button', () => {
  it('should open EditTaskModal on click', () => {
    wrapper.find('div').at(1).prop('onClick')();

    expect(wrapper.find('Connect(EditTaskModal)').prop('isOpen')).toBe(false);

    wrapper.find('button').at(0).prop('onClick')({ stopPropagation: () => {} });

    expect(wrapper.find('Connect(EditTaskModal)').prop('isOpen')).toBe(true);
  });
});

describe('remove task button', () => {
  it('should call startRemovePersonalTask on click', () => {
    wrapper.find('div').at(1).prop('onClick')();

    wrapper.find('button').at(1).prop('onClick')({ stopPropagation: () => {} });

    expect(startRemovePersonalTask).toHaveBeenCalledTimes(1);
    expect(startRemovePersonalTask).toHaveBeenLastCalledWith(id);
  });
});
