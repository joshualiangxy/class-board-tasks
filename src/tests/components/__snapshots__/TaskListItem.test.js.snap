// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clickable div should toggle visibility of div 1`] = `
<div>
  <input
    checked={false}
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    onClick={[Function]}
  >
    <h3>
      Watch lecture
    </h3>
    <h3
      className="after-deadline"
    >
      5th Jun 2020
    </h3>
    <div>
      <p>
        Catch up on lectures!
      </p>
      <button
        onClick={[Function]}
      >
        Edit Task
      </button>
      <button
        onClick={[Function]}
      >
        Remove Task
      </button>
    </div>
  </div>
  <Connect(EditTaskModal)
    groupModule="CS2030"
    isOpen={false}
    onRequestClose={[Function]}
    task={
      Object {
        "completed": false,
        "deadline": 1591329600000,
        "description": "Catch up on lectures!",
        "id": "5feb18c0-78a8-49b6-8584-75b8c1b40abc",
        "module": "CS2030",
        "title": "Watch lecture",
      }
    }
  />
</div>
`;

exports[`render should render TaskListItem 1`] = `
<div>
  <input
    checked={false}
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    onClick={[Function]}
  >
    <h3>
      Watch lecture
    </h3>
    <h3
      className="after-deadline"
    >
      5th Jun 2020
    </h3>
  </div>
  <Connect(EditTaskModal)
    groupModule="CS2030"
    isOpen={false}
    onRequestClose={[Function]}
    task={
      Object {
        "completed": false,
        "deadline": 1591329600000,
        "description": "Catch up on lectures!",
        "id": "5feb18c0-78a8-49b6-8584-75b8c1b40abc",
        "module": "CS2030",
        "title": "Watch lecture",
      }
    }
  />
</div>
`;

exports[`render should render TaskListItem with alt task 1`] = `
<div>
  <input
    checked={false}
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    onClick={[Function]}
  >
    <h3>
      Do lab
    </h3>
  </div>
  <Connect(EditTaskModal)
    groupModule="CS2030"
    isOpen={false}
    onRequestClose={[Function]}
    task={
      Object {
        "completed": false,
        "description": "",
        "id": "a1050220-fa88-4dc2-aa9d-9521ae95de47",
        "module": "CS2030",
        "title": "Do lab",
      }
    }
  />
</div>
`;
